feat: add the benchmark test for nydus

Desiki-high opened the pull request:
# To compare the container startup speed of nydus and oci, add the new feature.
## misc/benchmark/*.py is the benchmark workflow python script:
1. bench.py: bench the pull image time, create container time and start container time
2. benchmark.py : the benchmark satrt with four mode(oci, nydus without prefetch , nydus with prefetch all, nydus with prefetch file list)
3. convert.py: convet the image from oci to nydus by nydusify
4. metrics.py: use the nydus metrics api to collet the files which accessd by container in start time
5. prefetch_list.py: sort the file list by access time, maybe the more smart strategy will add in the future
6. util.py: the utility functions used in benchmark

## the other files in misc/benchmark is the config files or environment files:
1. cni_bridge.py: used by cni plugins
2. nydusd_config.json: nydusd config file
3. containerd_config.toml: snapshotter config file with containerd
4. nydus-snapshotter.service: systemd service to creat nydusd process

## add four jobs in smoke.yml :
### we use wondershape to limit to local net bandwith(10M in [config.yml](https://github.com/dragonflyoss/image-service/compare/master...Desiki-high:benchmark?expand=1#diff-924ec53aea92a71d8b43a94cc3e5a1260b1be0cbc934fcd21fbb4fc2b65057be))
1. env.IMAGE: choose the image for benchmark(if you modify this please also modiy the config.yaml)
2. benchmark-oci: benchmark for oci
3. benchmark-nydus-no-prefetch: benchmark for nydsu image (without prefetch)
4. benchmark-nydus-all-prefetch: benchmark for nydsu image (prefetch /)
5. benchmark-nydus-filelist-prefetch: benchmark for nydsu image (prefetch the file list create by prefetch_list.py )
### upload the benchmark result in the artifacts

jiangliu commented on misc/benchmark/config.json:
```diff
@@ -0,0 +1,30 @@
+{
```
how about renaming it as `nydus_config.json`?

jiangliu replied:
better use `nydusd_config.json`.

jiangliu commented on misc/benchmark/config.toml:
```diff
@@ -0,0 +1,9 @@
+[proxy_plugins]
```
Hoa about renaming it as `containerd_config.toml`?

jiangliu commented on misc/benchmark/bridge.conf:
```diff
@@ -0,0 +1,15 @@
+{
```
How about renaming it as `cni_bridge.conf`?

jiangliu commented:
It would be great to run performance test as a per-week cron task too with:
1) more container images
2) different network bandwidth configurations.

Desiki-high commented:
> In addition to run it in smoke tests, tt would be great to also run the performance test as a per-week cron task too with:
> 
> 1. more container images
> 2. different network bandwidth configurations.

In smoke tests, only one container image was tested due to time constraints. If you want to support more container images and different network bandwidth configurations, this script cant't support this idea currently.

imeoer commented on .github/workflows/smoke.yml:
```diff
@@ -128,6 +129,211 @@ jobs:
+      uses: actions/upload-artifact@v3
+      with:
+        name: benchmark-oci
+        path: misc/benchmark/${{env.IMAGE}}.csv
```
Should be `misc/benchmark-oci/${{env.IMAGE}}.csv` ?

Desiki-high replied:
Each test generates ${{env.IMAGE}}.csv in the current directory, the upload to the artifac. So  misc/benchmark-oci/${{env.IMAGE}}.csv is not better for smoke test. How about change to misc/benchmark/oci/${{env.IMAGE}}.csv? It will work better in [https://github.com/dragonflyoss/image-service/pull/1219#issuecomment-1510574406](https://github.com/dragonflyoss/image-service/pull/1219#issuecomment-1510574406)

imeoer commented on .github/workflows/smoke.yml:
```diff
@@ -128,6 +129,211 @@ jobs:
+      run: |
+        sudo install -D -m 755 contrib/nydusify/cmd/nydusify /usr/local/bin
+        sudo install -D -m 755 target/release/nydusd target/release/nydus-image /usr/local/bin
+        wget https://github.com/containerd/nydus-snapshotter/releases/download/v0.7.3/nydus-snapshotter-v0.7.3-x86_64.tgz
```
Better to make the snapshotter version as a variable defined on top.

imeoer commented on .github/workflows/smoke.yml:
```diff
@@ -128,6 +129,211 @@ jobs:
+        wget https://github.com/containerd/nydus-snapshotter/releases/download/v0.7.3/nydus-snapshotter-v0.7.3-x86_64.tgz
+        tar zxvf nydus-snapshotter-v0.7.3-x86_64.tgz
+        sudo install -D -m 755 nydus-snapshotter/containerd-nydus-grpc /usr/local/bin/
+        sudo wget https://github.com/containerd/nerdctl/releases/download/v1.3.0/nerdctl-1.3.0-linux-amd64.tar.gz
```
Ditto, make the version a variable.

imeoer commented on .github/workflows/smoke.yml:
```diff
@@ -128,6 +129,211 @@ jobs:
+  benchmark-nydus-all-prefetch:
+    runs-on: ubuntu-latest
+    needs: [contrib-build, nydus-build]
+    steps:
```
Can we reuse these steps `Download Nydus`, `Download Nydusify`, `Prepare Nydus Environment`, `Prepare Nydus Registry Environment`?

Desiki-high replied:
Actually, I tried it but I failed. Is any example can I learn from.

imeoer replied:
A possible approach is adding these as an individual job and upload these binary components and configs as an directory artifact, then download them on each benchmark job, but it doesn't work for `Prepare Nydus Registry Environment`.

imeoer commented on misc/benchmark/cni_bridge.conf:
```diff
@@ -0,0 +1,15 @@
+            { "dst": "0.0.0.0/0" }
+        ]
+    }
+}
```
The blank line should be added.

imeoer commented on misc/benchmark/config.yml:
```diff
@@ -0,0 +1,6 @@
+local_registry: localhost:5000
+insecure_local_registry: True
+bandwith: 81920
+image: wordpress:latest
```
The blank line should be added.

imeoer commented on misc/benchmark/nydus-snapshotter.service:
```diff
@@ -0,0 +1,18 @@
+StandardError=journal
+
+[Install]
+WantedBy=multi-user.target
```
The blank line should be added.

imeoer commented on misc/benchmark/nydusd_config.json:
```diff
@@ -0,0 +1,30 @@
+    },
+    "mode": "direct",
+    "digest_validate": false,
+    "iostats_files": true,
```
`iostats_files`, `access_pattern`, `latest_read_files`

These would affect runtime performance, have we only enabled these in the access file list analysis and disable them in our benchmark tests?

Desiki-high replied:
> `iostats_files`, `access_pattern`, `latest_read_files`
> 
> These would affect runtime performance, have we only enabled these in the access file list analysis and disable them in our benchmark tests?

Yes, we can move these configs to the runtime, it will work when we need.

imeoer commented:
Some other suggestions:

- Add `all-prefetch` test for nydus [zran](https://github.com/dragonflyoss/image-service/blob/master/docs/nydus-zran.md) image;
- Show the reports in the PR comment for the 5 benchmark tests;
- Compare the current benchmark tests with the master branch's like (It's just an example, don't worry about the specific values.):

**wordpress:latest**

| image type                | image size          | nydusify conversion | nerdctl run    | read data          |
| ------------------------- | ------------------- | ------------------- | -------------- | ------------------ |
| oci v1                    | 353.05MB            | -                   | 126s           | 353.05MB           |
| nydus no prefetch         | 337.94MB            | 29s (**+10%**)      | 11s (**-10%**) | 21.18MB            |
| nydus all prefetch        | 337.94MB (**+20%**) | 11s                 | 15s            | 28.78MB (**-30%**) |
| nydus filelist prefetch   | 337.94MB            | 11s (**+10%**)      | 15s (**-10%**) | 28.78MB            |
| nydus zran + all prefetch | 14MB                | 11s                 | 15s            | 28.78MB (**+10%**) |

**pytorch:latest**
...

Desiki-high commented:
> Some other suggestions:
> 
> * Add `all-prefetch` test for nydus [zran](https://github.com/dragonflyoss/image-service/blob/master/docs/nydus-zran.md) image;
> * Show the reports in the PR comment for the 5 benchmark tests;
> * Compare the current benchmark tests with the master branch's like (It's just an example, don't worry about the specific values.):
> 
> **wordpress:latest**
> 
> image type	image size	nydusify conversion	nerdctl run	read data
> oci v1	353.05MB	-	126s	353.05MB
> nydus no prefetch	337.94MB	29s (**+10%**)	11s (**-10%**)	21.18MB
> nydus all prefetch	337.94MB (**+20%**)	11s	15s	28.78MB (**-30%**)
> nydus filelist prefetch	337.94MB	11s (**+10%**)	15s (**-10%**)	28.78MB
> nydus zran + all prefetch	14MB	11s	15s	28.78MB (**+10%**)
> **pytorch:latest** ...

1. I never try the zran image, so maybe I need some time to work with that
2. Is there any action I can use to report the benchmark result in the PR comment? Actually it just like the https://github.com/dragonflyoss/image-service/pull/1219#discussion_r1168099340, I failed to do that.

adamqqqplay commented:
Thanks for your work, looks good overall.
Just a small suggestion, you could change your git email to a more official one. For example, Tencent QQ mailbox provides a domain name of foxmail.com.

Desiki-high commented:
> Thanks for your work, looks good overall. Just a small suggestion, you could change your git email to a more official one. For example, Tencent QQ mailbox provides a domain name of foxmail.com.

Thanks for you advice.

Desiki-high commented on .github/workflows/smoke.yml:
```diff
@@ -4,7 +4,7 @@ on:
     branches: ["*", "**"]
     paths-ignore: [ '**.md', '**.png', '**.jpg', '**.svg', '**/docs/**' ]
-  pull_request:
+  pull_request_target:
```
Is that don't triggered?

sctb512 commented on misc/benchmark/bench.py:
```diff
@@ -0,0 +1,259 @@
+
+class BenchRunner:
+    CMD_URL_WAIT = {
+        "wordpress": RunArgs(waitURL="http://localhost:80"),
```
Is it possible to load this info from a file so we can add more similar cases easily?

Desiki-high replied:
Yes, we can do that. Actually in this version, we only benchmark for wordpress. You can see https://github.com/dragonflyoss/image-service/issues/1220, we will add more images and configs to enhancement this feature.

sctb512 replied:
Got it.

Desiki-high committed:
add the benchmark test for nydus image

1. add the benchmark scripts in misc/benchmark
2. add five benchmark jobs in smoke test and the benchmark-result job for show the benchmark result in the PR comment

Signed-off-by: Desiki-high <ding_yadong@foxmail.com>

Desiki-high committed:
change the comment content

Signed-off-by: Desiki-high <ding_yadong@foxmail.com>

imeoer reviewed (approved):
LGTM
